pkg/backends/newrelic/flush.go

```diff
@@ -0,0 +1,110 @@
+	cb               func(*timeSeries)
+}
+
+// timeSeries represents a time series data structure.
+type timeSeries struct {
+	Metrics []interface{} `json:"metrics"`
+}
+
+// addMetric adds a metric to the series.
+func (f *flush) addMetric(n *Client, metricType string, value float64, persecond float64, hostname string, tags gostatsd.Tags, name string, timestamp gostatsd.Nanotime) {
+	standardMetric := setDefaultMetricSet(n, f, name, metricType, value, tags, timestamp)
+	if metricType == "counter" {
+		standardMetric[n.metricPrefix+n.metricPerSecond] = persecond
+	}
+	f.ts.Metrics = append(f.ts.Metrics, standardMetric)
+}
+
+// addMetric adds a timer metric to the series.
+func (f *flush) addTimerMetric(n *Client, metricType string, timer gostatsd.Timer, tagsKey, name string) {
+	timerMetric := setDefaultMetricSet(n, f, name, metricType, float64(timer.Count), timer.Tags, timer.Timestamp)
+
+	if !n.disabledSubtypes.Lower {
+		timerMetric[n.metricPrefix+n.timerMin] = timer.Min
+	}
+	if !n.disabledSubtypes.Upper {
+		timerMetric[n.metricPrefix+n.timerMax] = timer.Max
+	}
+	if !n.disabledSubtypes.Count {
+		timerMetric[n.metricPrefix+n.timerCount] = float64(timer.Count)
+	}
+	if !n.disabledSubtypes.CountPerSecond {
+		timerMetric[n.metricPrefix+n.metricPerSecond] = timer.PerSecond
+	}
+	if !n.disabledSubtypes.Mean {
+		timerMetric[n.metricPrefix+n.timerMean] = timer.Mean
+	}
+	if !n.disabledSubtypes.Median {
+		timerMetric[n.metricPrefix+n.timerMedian] = timer.Median
+	}
+	if !n.disabledSubtypes.StdDev {
+		timerMetric[n.metricPrefix+n.timerStdDev] = timer.StdDev
+	}
+	if !n.disabledSubtypes.Sum {
+		timerMetric[n.metricPrefix+n.timerSum] = timer.Sum
+	}
+	if !n.disabledSubtypes.SumSquares {
+		timerMetric[n.metricPrefix+n.timerSumSquares] = timer.SumSquares
+	}
+	for _, pct := range timer.Percentiles {
+		timerMetric[n.metricPrefix+pct.Str] = pct.Float
+	}
+	f.ts.Metrics = append(f.ts.Metrics, timerMetric)
+}
+
+func (f *flush) maybeFlush() {
+	if uint(len(f.ts.Metrics))+20 >= f.metricsPerBatch { // flush before it reaches max size and grows the slice
+		f.cb(f.ts)
+		f.ts = &timeSeries{
+			Metrics: make([]interface{}, 0, f.metricsPerBatch),
+		}
+	}
+}
+
+func (f *flush) finish() {
+	if len(f.ts.Metrics) > 0 {
+		f.cb(f.ts)
+	}
+}
+
+func setDefaultMetricSet(n *Client, f *flush, metricName, Type string, Value float64, tags gostatsd.Tags, timestamp gostatsd.Nanotime) map[string]interface{} {
+	defaultMetricSet := map[string]interface{}{}
+	defaultMetricSet["interval"] = f.flushIntervalSec
+	defaultMetricSet["timestamp"] = timestamp
+	defaultMetricSet["event_type"] = n.eventType
+	defaultMetricSet["integration_version"] = integrationVersion
+	defaultMetricSet["protocol_version"] = protocolVersion
+
+	defaultMetricSet[n.metricPrefix+n.metricType] = Type
+	defaultMetricSet[n.metricPrefix+n.metricName] = metricName
+	defaultMetricSet[n.metricPrefix+n.metricValue] = Value
+
+	for _, tag := range tags {
+		if tag != "" && strings.Contains(tag, ":") {
```

Comment:
This will drop a valueless tag.  I would favor handling it like ```tag:valueless``` or something along those lines, otherwise, consider the following two metrics:

- metric[foo:bar, baz] = 5
- metric[foo:bar] = 10

That would emit two metrics with foo=bar, possibly leading to confusion.

Reply:
Single tags, will now come as { "tag": "true" }. Decided on "true" instead as this is used in other New Relic events, such as processSamples where the same is done with attributes that just exist eg. "contained" to indicate if it is containerized.